### PR TITLE
Linelist: strongest_transitions() and available_transitions() implemented

### PR DESCRIPTION
**Pull request by ntejos**
strongest_transitions() and available_transitions() implemented in LineList.

### CODE REVIEW

File: linetools/lists/linelist.py [hunk 1]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+        elif isinstance(wvlims, Quantity): # or quantity
+            pass
+        else:
+            raise ValueError('wvlims has to be tuple or Quantity')
+        if len(wvlims) != 2:
+            raise SyntaxError('wlims has to be of size len()== 2; Please correct format')
+        if wvlims[0] >= wvlims[1]:
+            raise ValueError('Minimum limit (wlims[0]) is not smaller than maximum limit (wlims[1]); please correct')
+
+        if n_max < 1:
```
Comment (profxj) [inline linetools/lists/linelist.py]: Should you query whether n_max is an int?

File: linetools/lists/linelist.py [hunk 2]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+
+
+        Parameters:
+        ----------
+        line: str or Quantity
+            Name of line. (e.g. 'HI 1215', 'HI', 'CIII', 'SiII', 1215.6700*u.AA)
+            [Note: when string contains spaces it only considers the first
+             part of it, so 'HI' and 'HI 1215' and 'HI 1025' are all equivalent]
+            [Note: to retrieve an unknown line use string 'unknown']
+        wvlims : tuple of Quantity
```
Comment (profxj) [inline linetools/lists/linelist.py]: Why have this be in observed space?
I think the input should be in rest wavelength.
Leave it to the user to divide by (1+z)

File: linetools/lists/linelist.py [hunk 3]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+            return data
+        elif np.sum(cond) == 1: #only 1 case from a QTable format
+            name = data[cond]['name'][0]
+            return self.__getitem__(name)
+        else:
+            #remove transitions out of range
+            data = data[cond]
+            #add strength column to data
+            strength = data['wrest'] * data['f']
+            data.add_column(Column(data=strength,name='strength'))
```
Comment (profxj) [inline linetools/lists/linelist.py]: While this surely works, I think you can slice the Table, i.e. provide it a sorted list of indices.

File: linetools/lists/linelist.py [hunk 4]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+            Redshift
+        n_max : int, optional
+            Maximum number of transitions retrieved
+        n_max_tuple : int, optional
+            Maximum number of transitions in a given ion species to 
+            retrieve. e.g., if Lyman series are all available, it will 
+            retrieve only up to Lyman gamma if n_max_tuple=3.
+        min_strength : float, optional
+            Minimum strenght calculated from log10(wrest * fosc * abundance)
+            In thin space HI 1215 has 14.7 [not]
```
Comment (profxj) [inline linetools/lists/linelist.py]: Am confused by this line..

File: linetools/lists/linelist.py [hunk 5]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+            raise SyntaxError('Both n_max and n_max_tuple must be integers!')
+
+        if n_max == 0:
+            return None
+
+        data = copy.deepcopy(self._data)
+
+        #identify unique ion_names (e.g. HI, CIV, CIII)
+        unique_ion_names = list(set([name.split(' ')[0] for name in data['name']]))
+        unique_ion_names = np.array(unique_ion_names)
```
Comment (profxj) [inline linetools/lists/linelist.py]: By 'unique', do you mean non-duplicated?  Seems like this array will have duplicates.  You might use np.unique

File: linetools/lists/linelist.py [hunk 6]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+    def available_transitions(self, wvlims, z, n_max=100,n_max_tuple=3, min_strength=4):
+        """For a given wavelength range in Angstroms, wvlims=(wv_min,wv_max),
+        this function retrieves the n_max_tuple strongest transitions per each ion 
+        species in the LineList available at redshift z within such a wavelength 
+        range. The output is sorted by strength of the strongest available transition 
+        per ion species, modualted by abundance.
+
+        Parameters:
+        ----------
+        wvlims : tuple of Quantity
```
Comment (profxj) [inline linetools/lists/linelist.py]: Again, I'd just input wrest range

File: linetools/lists/linelist.py [hunk 7]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+            #no matches
+            return None
+
+        #create Table
+        unique = Table()
+        unique.add_column(Column(data=ion_name,name='name'))
+        unique.add_column(Column(data=strength,name='strength'))
+
+        #sort by strength
+        unique.sort(['strength'])
```
Comment (profxj) [inline linetools/lists/linelist.py]: Why the [] around 'strength' ?

File: linetools/lists/linelist.py [hunk 8]
```diff
@@ -371,7 +564,8 @@ def __getitem__(self,k, tol=1e-3*u.AA):
 
         Returns:
         ----------
-        dict (from row in the data table) or Table (tuple)
+        dict (from row in the data table if only 1 line is found) or QTable (tuple
```
Comment (profxj) [inline linetools/lists/linelist.py]: Why 'tuple'?

File: linetools/lists/linelist.py [hunk 9]
```diff
@@ -344,7 +349,195 @@ def all_transitions(self,line):
...
+
+        Parameters:
+        ----------
+        wvlims : tuple of Quantity
+            Wavelength range, e.g. wvlims=(1100*u.AA, 3200*u.AA)
+        z : float
+            Redshift
+        n_max : int, optional
+            Maximum number of transitions retrieved
+        n_max_tuple : int, optional
```
Comment (profxj) [inline linetools/lists/linelist.py]: Does this mean you don't compare against ions?  You take the strongest 3 of each?  Am a bit confused on this one.

File: linetools/lists/linelist.py (hunk 1 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: Definitively.

File: linetools/lists/linelist.py (hunk 4 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: The idea is to provide a minimum strength so the list of lines is reduced, where strength is proportional to: (wrest * fosc * abundance), with abundance in log10 space (i.e. abundance_H=12.0).

File: linetools/lists/linelist.py (hunk 5 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: Will give it a try.

File: linetools/lists/linelist.py (hunk 7 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: Is not strictly needed but Table.sort() will convert the input to a list anyways in cases where input is not list already... so a if() statement is avoided.

File: linetools/lists/linelist.py (hunk 8 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: I think you wrote the original documentation; here tuple refers to the input 'k'

File: linetools/lists/linelist.py (hunk 9 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: The idea is to focus in the strongest available transition for a given ion species, and then look up to n_max_tuple transitions of that same species independently of its strength compared to others. This is helpful for the line identification process, where finding evidence of doublets (or tuplets) adds considerably diagnostic power. We can, of course, discuss further what these functions should really do.

File: linetools/lists/linelist.py (hunk 3 above)
Comment (ntejos) [inline linetools/lists/linelist.py]: Will give it a try.

**Comment by profxj** [comment]
Merging.

**Pull request merged**